### PR TITLE
feat: add initial implementation

### PR DESCRIPTION
**Pull request by onebytegone**

### CODE REVIEW

File: src/list-indentation.js [hunk 1]
```diff
@@ -1,14 +1,192 @@
...
+
+               expectedIndent = token.startColumn + child.endColumn - child.startColumn - 1;
+               skipTo('lineEnding');
+               return;
+            }
+
+            // When `ul_indent` is configured, allow sub-items to be indented at a
+            // different depth. This is to support behavior found in ul-indent (MD007).
+            // See: https://github.com/DavidAnson/markdownlint/blob/main/doc/md007.md
+            if (params.config.ul_indent && token.type === 'listUnordered' && [ 'listOrdered', 'listUnordered' ].includes(child.type)) {
```
Comment (MrMarCode) [inline src/list-indentation.js:156]: Although this is primarily here to support MD007, would it make sense for the setting to be adjusted so that it's more broad and included any sub items (blockquote, fenced code, paragraphs, etc...)? ie:
```js
await testValidExample([
   '* first-level list item',
   '',
   '   > blockquote',
], { 'ul_indent': 3 });
```

Personally, I like the way the rule currently works. But from another point of view, a developer might see this setting as causing inconsistency across columns. For example, if we set ul-indent to 3 under the current rule, the following is completely valid:
```md
* first-level list item
  continued paragraph 2 indent.
   * sub ul with 3 indent.
  > blockquote with 2 indent
```

Do you think something like this would look better?
```
* first-level list item
  continued paragraph 2 indent.
   * sub ul with 3 indent.
   > blockquote with 3 indent
```

Or even this?
```
* first-level list item
   continued paragraph 3 indent.
   * sub ul with 3 indent.
   > blockquote with 3 indent
```

File: tests/list-indentation.test.js [hunk 2]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '    * ul item',
+         '      continued paragraph',
+         '  * ul item',
+         '    continued paragraph',
+         '  1. ol item',
+         '     continued paragraph',
+      ]);
+   });
+
+   it('properly validates indented paragraphs with blockquotes', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: This is a leftover test I had included but I think it can be removed. Other tests seem to test the various edge cases in this test (see 'reports no errors for correct indention of item under long ol list item prefixes' and 'reports no errors in the grab bag test')

File: tests/list-indentation.test.js [hunk 3]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '    Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
+         'Pellentesque vestibulum lectus non tellus congue,',
+         '  eu ultricies metus ultrices.',
+         '',
+         'Curabitur ut posuere est, quis convallis orci.',
+         '   Nam lectus magna, pellentesque at rutrum in, commodo a nisi.',
+      ]);
+   });
+
+   it('properly validates indented paragraphs', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Can be removed. Tested by 'reports no errors in the grab bag test'

File: tests/list-indentation.test.js [hunk 4]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '     ```',
+         '                  This content is not checked();',
+         '             not checked();',
+         '            not checked();',
+         '             not checked();',
+         '             not checked();',
+      ]);
+   });
+
+   it('should validate ol inside of a ul with no starting indentation (4 layers deep)', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Could possibly removed or moved inside of 'reports no errors for left-aligned list items'

File: tests/list-indentation.test.js [hunk 5]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+      await testInvalidExample([
+         '* ul item',
+         '   * ul item',
+         '       1. ol item',
+         '          * ul item',
+         '       2. ol item',
+      ], { 'ul_indent': 3 });
+   });
+
+   it('should fail short indentation for a sub-list', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports errors for incorrectly aligned blocks'

File: tests/list-indentation.test.js [hunk 6]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+   });
+
+   it('should fail short indentation for a blockquote', async () => {
+      await testInvalidExample([
+         '>> ul item',
+         '   >> ul item',
+      ]);
+   });
+
+   it('should fail long indentation for an ol inside of an ul', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports errors for incorrectly aligned blocks'

File: tests/list-indentation.test.js [hunk 7]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+      await testValidExample([
+         '### 1. ol item',
+         '   1. ol item',
+         '      * ul item',
+         '   2. ol item',
+         '## 2. ol item',
+      ]);
+   });
+
+   it('should validate no lists', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports no errors for left-aligned paragraphs'

File: tests/list-indentation.test.js [hunk 8]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '   * **Message:**',
+         '',
+         '     > a\\',
+         '     > b',
+         '     > c',
+         '',
+      ]);
+   });
+
+   it('properly validates indented code blocks', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports no errors for properly aligned code blocks'

File: tests/list-indentation.test.js [hunk 9]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '```',
+         '              This content is not checked();',
+         '          not checked();',
+         '         not checked();',
+         '      not checked();',
+         '```',
+      ]);
+   });
+
+   it('validates 3 space indent', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports no errors for indented lists'

File: tests/list-indentation.test.js [hunk 10]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+      await testValidExample([
+         '',
+         '## Legal Offices Page',
+         '',
+         '   * Includes body content and a [**Contact Information Chooser**](#TODO) for legal office',
+         '     contact information',
+      ]);
+   });
+
+   it('properly validates starting fenced code', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports no errors for properly aligned code blocks'

File: tests/list-indentation.test.js [hunk 11]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+
+   it('should validate list items inside of a blockquote', async () => {
+      await testValidExample([
+         '>> * ul item',
+         '>> * ul item',
+      ]);
+   });
+
+   it('should fail short indentation for a blockquote', async () => {
+      await testInvalidExample([
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: This test could be moved to 'reports errors for improperly aligned blockquotes'

File: tests/list-indentation.test.js [hunk 12]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+
+   it('reports no errors for properly aligned links in lists', async () => {
+      await testValidExample([
+         '1. [This is a multiline link',
+         '   to test](#test-link)',
+         '2. [another link](#another-link)',
+      ]);
+   });
+
+   it('should validate list items inside of a blockquote', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports no errors in the grab bag test'

File: tests/list-indentation.test.js [hunk 13]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+   });
+
+   it('stars (bold or italics) should not throw off indent check', async () => {
+      await testValidExample([
+         '## test page',
+         '',
+         '   1. contact information',
+         '      1. contact information',
+         '         * contact information',
+         '            * Includes body content and a [**Contact Information Chooser**](#TODO) test',
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: This section of the test could be moved to 'reports no errors in the grab bag test' or we could probably remove it since we are no longer using regex and this error probably doesn't have to be worried about anymore.

File: tests/list-indentation.test.js [hunk 14]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+         '               * contact information',
+      ], { 'ul_indent': 3 });
+   });
+
+   it('should prioritize indent config setting over dynamic indent for child unordered lists', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: Seems to be handled by 'reports errors for non-left-aligned list items'

File: tests/list-indentation.test.js [hunk 15]
```diff
@@ -7,24 +7,1216 @@ const { markdownlint } = require('markdownlint').promises,
...
+      await testValidExample([
+         '   1. ol item',
+         '      1. ol item',
+         '         * ul item',
+         '            * ul item',
+         '               * ul item',
+      ], { 'ul_indent': 3 });
+   });
+
+   it('should validate ol inside of a ul with starting indentation of 3 (3 layers deep)', async () => {
```
Comment (MrMarCode) [inline tests/list-indentation.test.js]: This test could be moved inside of 'reports no errors for left-aligned list items'

File: src/list-indentation.js (hunk 1 above)
Comment (onebytegone) [reply src/list-indentation.js:156]: Good question! For now, I think we leave the behavior as is. The current behavior can be summarized as:

> Any list items under an unordered list should be indented to `ul_indent`. The "content" of the ul list item should be left aligned.

This lends towards something like what listed below. All the `a` content is indented a the same level and the sub list items (`b` and `c`) obey `ul_indent`.

~~~text
* aaaaa aaaaa aaaaa aaa aaaa, 
  aaaaaaaaaaa aaaaaaaaaa aaaa.

   * bbbbb bbbbbbbb bbbbbbbb bbbbb bbbbb.

  > aaaaaaaa aaaaa aaaaaaaa aaaa

  aaaaaaaa aaaaa aaaaaaaa.

   1. ccc ccccc ccc cccc ccccc.

  ```
  aaaaaaaa aaaaa aaaaaaaa.
  ```
~~~

**Review by MrMarCode — APPROVED** [review]
LGTM :slightly_smiling_face::+1:
I suppose now all that's left is changing the name of the rule.
